[improve][doc] Improve golang client doc (#17455)
apache/pulsar: site2/docs/client-libraries-go.md

defer consumer.Close()

for i := 0; i < 10; i++ {
    // may block here
	msg, err := consumer.Receive(context.Background())
	if err != nil {

> 		log.Fatal(err)

Reviewer: @labuladong generally Golang use tab for indent. In this patch we mix tabs and spaces, you may keep them consistent.

Reviewer: Not really. [These lines](https://github.com/apache/pulsar/blob/a053323e23998d9b7e746933cbce43590ad956b9/site2/docs/client-libraries-go.md?plain=1#L71-L76) are still using tabs. I suggest you replace all leading space in Golang code blocks with tabs.

Author: Code in markdown block will treat `\t` as 4 spaces. But you are right, I can use regex to replace them.

Reviewer: Yes. I'm OK to use spaces here since Go formatter should be able to format it. My suggestion focuses on consistently - that is, we don't use both.

### Consumer operations

All available methods of `Consumer` interface are [here](https://pkg.go.dev/github.com/apache/pulsar-client-go/pulsar#Consumer).

#### Create single-topic consumer

```go
client, err := pulsar.NewClient(pulsar.ClientOptions{URL: "pulsar://localhost:6650"})
if err != nil {
    log.Fatal(err)
}

defer client.Close()

consumer, err := client.Subscribe(pulsar.ConsumerOptions{
    // fill `Topic` field will create a single-topic consumer
    Topic:            "topic-1",
    SubscriptionName: "my-sub",
    Type:             pulsar.Shared,
})
if err != nil {
    log.Fatal(err)
}
defer consumer.Close()
```

#### Create regex-topic consumer

```go
client, err := pulsar.NewClient(pulsar.ClientOptions{
		URL: "pulsar://localhost:6650",
})
defer client.Close()

topicsPattern := "persistent://public/default/topic.*"
opts := pulsar.ConsumerOptions{
    // fill `TopicsPattern` field will create a regex consumer
    TopicsPattern:    topicsPattern,
    SubscriptionName: "regex-sub",
}

consumer, err := client.Subscribe(opts)
if err != nil {
    log.Fatal(err)
}
defer consumer.Close()
```

> #### Create multi-topic Consumer

Reviewer 2: ```suggestion
#### Create multi-topic consumer
```

#### Create single-topic consumer

```go
client, err := pulsar.NewClient(pulsar.ClientOptions{URL: "pulsar://localhost:6650"})
if err != nil {
    log.Fatal(err)
}

defer client.Close()

consumer, err := client.Subscribe(pulsar.ConsumerOptions{
    // fill `Topic` field will create a single-topic consumer
    Topic:            "topic-1",
    SubscriptionName: "my-sub",
    Type:             pulsar.Shared,
})
if err != nil {
    log.Fatal(err)
}
defer consumer.Close()
```

#### Create specific-partition consumer

```go
client, err := pulsar.NewClient(pulsar.ClientOptions{URL: "pulsar://localhost:6650"})
if err != nil {
    log.Fatal(err)
}

defer client.Close()

// subscribe a specific partition of a topic
consumer, err := client.Subscribe(pulsar.ConsumerOptions{
    // pulsar partition is a special topic has the suffix '-partition-xx'
    Topic:            "my-partitioned-topic-partition-2",
    SubscriptionName: "my-sub",
    Type:             pulsar.Shared,

> })

Reviewer 3: Why do we add this example? I think subscribing to a specific partition is not a good practice.

Author: This example splits from here: 
https://pulsar.apache.org/docs/next/client-libraries-go#how-to-use-message-router-in-producer

Reviewer 3: The original example is to introduce how to use the message router. It's from the producer part and is just used to test the custom message router.

Author: You are right, I'll remove this example.

// create consumer
consumer, err := client.Subscribe(pulsar.ConsumerOptions{
    Topic:            topic,
    SubscriptionName: "my-sub1",
    Type:             pulsar.Shared,
})
if err != nil {
    log.Fatal(err)
}
defer consumer.Close()

// receive message with a timeout
msg, err := consumer.Receive(ctx)
if err != nil {
    log.Fatal(err)
}
fmt.Println(msg.Payload())
```

> #### How to use schema in consumer

Reviewer 2: ```suggestion
#### Use schema in consumer
```

	defer client.Close()

	// we can listen this channel
	channel := make(chan pulsar.ConsumerMessage, 100)

	options := pulsar.ConsumerOptions{
		Topic:            "topic-1",
		SubscriptionName: "my-subscription",
		Type:             pulsar.Shared,
		// fill `MessageChannel` field will create a listener
		MessageChannel: channel,
	}

	consumer, err := client.Subscribe(options)
	if err != nil {
		log.Fatal(err)
	}

	defer consumer.Close()

	// Receive messages from channel. The channel returns a struct `ConsumerMessage` which contains message and the consumer from where
	// the message was received. It's not necessary here since we have 1 single consumer, but the channel could be
	// shared across multiple consumers as well
	for cm := range channel {
		consumer := cm.Consumer
		msg := cm.Message
		fmt.Printf("Consumer %s received a message, msgId: %v, content: '%s'\n",
			consumer.Name(), msg.ID(), string(msg.Payload()))

		consumer.Ack(msg)
	}
}
```

#### Receive message with timeout

```go

> client, err := NewClient(pulsar.ClientOptions{

Reviewer 4: ```suggestion
client, err := pulsar.NewClient(pulsar.ClientOptions{
```

	ID   int    `json:"id"`
	Name string `json:"name"`
}

var (
	exampleSchemaDef = "{\"type\":\"record\",\"name\":\"Example\",\"namespace\":\"test\"," +
		"\"fields\":[{\"name\":\"ID\",\"type\":\"int\"},{\"name\":\"Name\",\"type\":\"string\"}]}"
)

> client, err := NewClient(pulsar.ClientOptions{

Reviewer 4: ```suggestion
client, err := pulsar.NewClient(pulsar.ClientOptions{
```

Reviewer 4: `consumerJS := NewJSONSchema(exampleSchemaDef, nil)` -> `consumerJS := pulsar.NewJSONSchema(exampleSchemaDef, nil) `

  scrape_interval: 10s
  static_configs:
  - targets:
  - localhost: 2112
```

Now you can query Pulsar client metrics on Prometheus.

### Consumer configuration

All available options of `ConsumerOptions` are [here](https://pkg.go.dev/github.com/apache/pulsar-client-go/pulsar#ConsumerOptions).

## Readers

Pulsar readers process messages from Pulsar topics. Readers are different from consumers because with readers you need to explicitly specify which message in the stream you want to begin with (consumers, on the other hand, automatically begin with the most recent unacked message). You can [configure](#reader-configuration) Go readers using a `ReaderOptions` object. Here's an example:

```go
reader, err := client.CreateReader(pulsar.ReaderOptions{
    Topic:          "topic-1",
    StartMessageID: pulsar.EarliestMessageID(),
})
if err != nil {
    log.Fatal(err)
}
defer reader.Close()
```

### Reader operations

> All available method of `Reader` interface are [here](https://pkg.go.dev/github.com/apache/pulsar-client-go/pulsar#Reader).

Reviewer 5: ```suggestion
All available methods of the `Reader` interface are [here](https://pkg.go.dev/github.com/apache/pulsar-client-go/pulsar#Reader).
```

> #### How to use reader to read specific message

Reviewer 2: ```suggestion
#### Use reader to read specific message
```

// create producer
producer, err := client.CreateProducer(pulsar.ProducerOptions{
    Topic:           topic,
    DisableBatching: true,
})
if err != nil {
    log.Fatal(err)
}
defer producer.Close()

// send 10 messages
msgIDs := [10]pulsar.MessageID{}
for i := 0; i < 10; i++ {

>     msgID, _ := producer.Send(ctx, &pulsar.ProducerMessage{

Reviewer 2: @nodece 
could you please review this PR from a technical perspective? Thank you!

Author: Seems that the original code snippet is copied from the test cases, which is unnecessary in the documentation.

Reviewer 4: Hi @Anonymitaet, I suggest we remove parameter details, and then use https://pkg.go.dev/github.com/apache/pulsar-client-go/pulsar instead, when the go-client updates, we need to update our documentation again, what do you think of this.

Author: I suggest reserving them because other clients' documentation also has parameter details.

Reviewer 3: +1 for removing them. It's hard to maintain the parameters detail in two places: the go doc and the pulsar doc website. Or we can provide some commonly used parameters in this doc and let's user jump to go doc for more detail.
In fact, the java client doc also encourages the user to jump to the java doc to get more details. But leave the loadConf configuration which is generated from the code in the doc.

Reviewer 2: @nodece I'm on your side. Referencing the source material reduces maintenance cost

@labuladong FYI: to reduce manual work, we (@SignorMercurio @urfreespace) are trying to generate client configurations docs from code automatically (https://github.com/apache/pulsar/pull/17198), which means we'll remove the [configuration instructions - table](https://pulsar.apache.org/docs/next/client-libraries-java#configure-producer) from client doc.

Author: Got it. I will remove the parameter details and link to https://pkg.go.dev/github.com/apache/pulsar-client-go/pulsar
